### PR TITLE
[app-metrics][iOS] Stop inventing symbols for native crash frames

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

We symbolicate crash frames on device with `dladdr`, which returns the nearest symbol at or below an address and never says how far away it was. Release binaries are stripped, so the symbols it has left can sit megabytes apart, and every frame in between gets named after whichever one happens to precede it.

Nothing marks those names as guesses, so they read as fact and send you after code that never ran. Here's a real trace from observe-tester, with the long frames elided:

```
Swift._assertionFailure(...) -> Swift.Never
Observe + 19440
ExpoModulesCore.ExpoFabricView.(previousProps in ...).modify : Swift.Dictionary<Swift.String, Any>
ExpoModulesCore.ExpoFabricView.(previousProps in ...).modify : Swift.Dictionary<Swift.String, Any>
ExpoModulesCore.ExpoFabricView.(previousProps in ...).modify : Swift.Dictionary<Swift.String, Any>
ExpoModulesCore.ExpoFabricView.(previousProps in ...).modify : Swift.Dictionary<Swift.String, Any>
ExpoModulesCore.ExpoFabricView.(previousProps in ...).modify : Swift.Dictionary<Swift.String, Any>
ExpoModulesCore.ExpoFabricView.(previousProps in ...).modify : Swift.Dictionary<Swift.String, Any>
closure #1 @ExpoModulesJSI.JavaScriptActor @Sendable () -> Swift.Bool in closure #1 (...) in call #1 (...) in ExpoModulesJSI.(createFunctionClosure in ...)(...)
generic not re-abstracted specialization <()> of implicit closure #1 ... in static ExpoModulesJSI.JavaScriptActor.assumeIsolated<A ...>(...)
@objc call #1 (...) -> Swift.Bool in ExpoModulesJSI.(createFunctionClosure in ...)(...)
std::__1::__invoke_result_impl<void, expo::createHostFunction(...)::'lambda'(...)&, ...>::type std::__1::__invoke[abi:nqe210106]<...>(...)
std::__1::function<facebook::jsi::Value (...)>::operator()(...) const
facebook::hermes::(anonymous namespace)::HermesRuntimeImpl::HFContext::func(void*, hermes::vm::Runtime&)
hermes::vm::NativeFunction::_nativeCall(hermes::vm::NativeFunction*, hermes::vm::Runtime&)
hermes::vm::Interpreter::handleCallSlowPath(hermes::vm::Runtime&, hermes::vm::PinnedHermesValue*)
… +27 more frames
```

Fabric prop diffing has no business on that call path, and it wasn't there: all six frames are unrelated code, each named after a symbol hundreds of kilobytes away. `renderFrame` dropped the binary and offset whenever a symbol resolved, so there was nothing in the trace to check them against either. The same crash now reads:

```
Swift._assertionFailure(...) -> Swift.Never (libswiftCore.dylib + 2330452)
Observe + 19440
Observe + 2080032
Observe + 2080136
Observe + 2761984
Observe + 2763188
Observe + 2764756
Observe + 2765556
closure #1 @ExpoModulesJSI.JavaScriptActor @Sendable () -> Swift.Bool in closure #1 (...) in call #1 (...) in ExpoModulesJSI.(createFunctionClosure in ...)(...) (ExpoModulesJSI + 113544)
generic not re-abstracted specialization <()> of implicit closure #1 ... in static ExpoModulesJSI.JavaScriptActor.assumeIsolated<A ...>(...) (ExpoModulesJSI + 77356)
@objc call #1 (...) -> Swift.Bool in ExpoModulesJSI.(createFunctionClosure in ...)(...) (ExpoModulesJSI + 113704)
std::__1::__invoke_result_impl<void, expo::createHostFunction(...)::'lambda'(...)&, ...>::type std::__1::__invoke[abi:nqe210106]<...>(...) (ExpoModulesJSI + 128468)
std::__1::function<facebook::jsi::Value (...)>::operator()(...) const (hermesvm + 183960)
facebook::hermes::(anonymous namespace)::HermesRuntimeImpl::HFContext::func(void*, hermes::vm::Runtime&) (hermesvm + 183248)
hermes::vm::NativeFunction::_nativeCall(hermes::vm::NativeFunction*, hermes::vm::Runtime&) (hermesvm + 1060644)
hermes::vm::Interpreter::handleCallSlowPath(hermes::vm::Runtime&, hermes::vm::PinnedHermesValue*) (hermesvm + 1103752)
… +27 more frames
```

# How

`resolveSymbol` now compares the address against `dli_saddr` and throws the match away when it lands too far past the symbol's start. There's no recovering the real name on device, so the frame falls back to `binaryName + offset`.

Two things keep that from losing what we already had. MetricKit ships symbols for frames it can attribute and we were overwriting them with our own nil, so those survive now. And every frame keeps its binary and offset even when a symbol resolved, so a suspicious name can be checked against the archive.

The frame limit goes to 50, since a crash that starts in JS spends most of its stack inside the Hermes interpreter and 25 was cutting traces in half. New attributes cover what the reports couldn't tell us: thread count and whether MetricKit attributed one, rendered against total frames, the UUIDs of the binaries named (an offset means nothing without the build), and the faulting memory region, which is the only thing separating a stack overflow from any other bad access.

On-device symbolication of app code is a dead end while the binary is stripped, and iOS 27 gives us a better path, so this is the fallback for older versions rather than something to build on.

# Test Plan

`et check-packages expo-app-metrics` and `et native-unit-tests -p ios --packages expo-app-metrics`.

The traces above are from observe-tester's debug screen on a TestFlight build, before and after. `libswiftCore.dylib`, `ExpoModulesJSI` and `hermesvm` still symbolicate, so the guard only drops what it can't name. Replaying it over the shipped `.ipa`'s symbol table rejects every misattribution those reports contained and still resolves all 984 addresses that hit a symbol exactly.

### CODE REVIEW

File: packages/expo-app-metrics/ios/CrashReporting/CrashReportSymbolicator.swift [hunk 1]
```diff
@@ -75,6 +84,19 @@ enum CrashReportSymbolicator {
...
     }
+    // A stripped binary can leave gaps megabytes wide between the symbols it keeps, and `dladdr`
+    // will happily name a frame after whichever one happens to precede it. Nothing marks that as a
+    // guess, so it sends you looking at code that never ran. Better to admit we don't know and let
+    // the caller fall back to `binaryName + offset`.
+    guard let symbolStart = info.dli_saddr else {
+      return nil
+    }
+    let symbolAddress = UInt64(UInt(bitPattern: symbolStart))
+    guard currentAddress >= symbolAddress,
```
Comment (jakub-agent) [inline packages/expo-app-metrics/ios/CrashReporting/CrashReportSymbolicator.swift]: [🏗️ **design**] This guard is the whole PR and there's no test for it. Can we pull the distance check out into a small function (address + symbol start in, `Bool` out) and unit test it? `resolveSymbol` needs a live `dladdr`, so nothing here is covered today.

File: packages/expo-app-metrics/ios/CrashReporting/CrashReportSymbolicator.swift [hunk 2]
```diff
@@ -25,6 +25,13 @@ import MachO
 /// Where `currentLoadAddress` is the slide-adjusted address of the binary's first segment
 /// (typically `__TEXT`) in this process. We grab it from the dyld image table.
 enum CrashReportSymbolicator {
+  /// How far past a symbol's start a frame can land and still be attributed to it.
+  ///
+  /// `dladdr` tells us the nearest symbol at or below an address, but not how far away it was, so
+  /// this is all we have to go on. It's big enough to cover almost every gap between the symbols
+  /// our release builds keep, and far smaller than the gaps that produced wrong names.
+  private static let maxSymbolMatchDistance: UInt64 = 8 * 1024
```
Comment (jakub-agent) [inline packages/expo-app-metrics/ios/CrashReporting/CrashReportSymbolicator.swift]: [🏗️ **design**] Where does 8 KB come from? The test plan says every address that "hit a symbol exactly" still resolves, but offset 0 doesn't exercise the window at all — a function larger than 8 KB (the Hermes interpreter loop, big template instantiations) would now lose a name that was correct.

If this 
> It's big enough to cover almost every gap between the symbols   
  /// our release builds keep, and far smaller than the gaps that produced wrong names.

is the explanantion maybe we can add `8KB is big enough......`

File: packages/expo-app-metrics/ios/CrashReporting/CrashReportSymbolicator.swift [hunk 3]
```diff
@@ -75,6 +84,19 @@ enum CrashReportSymbolicator {
     else {
       return nil
     }
+    // A stripped binary can leave gaps megabytes wide between the symbols it keeps, and `dladdr`
```
Comment (jakub-agent) [inline packages/expo-app-metrics/ios/CrashReporting/CrashReportSymbolicator.swift]: [🐧 nit] This repeats the `maxSymbolMatchDistance` doc comment almost word for word, ~55 lines up. Let's keep the reasoning in one place.

File: packages/expo-app-metrics/ios/CrashReporting/CrashReport.swift [hunk 4]
```diff
@@ -187,17 +217,49 @@ public struct CrashReport: Codable, Sendable {
...
-      return "\(binaryName) + \(offset)"
+    guard !uuidsByBinary.isEmpty else {
+      return nil
     }
-    if let address = frame.address {
-      return "0x\(String(address, radix: 16))"
+    return uuidsByBinary.sorted { $0.key < $1.key }.map { "\($0.key):\($0.value)" }
+  }
+
+  /// Renders one frame as `<symbol> (<binary> + <offset>)`.
```
Comment (jakub-agent) [inline packages/expo-app-metrics/ios/CrashReporting/CrashReport.swift:237]: [💡 suggestion] Two cases here render something that doesn't help: with no offset you get `frame0 (TestApp)`, and a MetricKit `<redacted>` symbol becomes `<redacted> (Observe + 2080032)`. WDYT about appending the location only when we actually have an offset or an address?

File: packages/expo-app-metrics/ios/CrashReporting/CrashReport.swift [hunk 5]
```diff
@@ -6,7 +6,12 @@
 /// payload time window. Unhandled JavaScript errors are recorded separately as `js.exception` log
 /// events (see `ErrorReport`).
 public struct CrashReport: Codable, Sendable {
-  private static let maxLogStackFrames = 25
+  /// Frames rendered into `exception.stacktrace`.
+  ///
+  /// A crash that started in JS spends most of its stack inside the Hermes interpreter before it
+  /// ever reaches our code, so a small limit cuts off the part you actually want. 50 gets us the
+  /// whole trace and still leaves plenty of room under `maxLogStacktraceLength`.
+  private static let maxLogStackFrames = 50
```
Comment (jakub-agent) [inline packages/expo-app-metrics/ios/CrashReporting/CrashReport.swift:13]: [🏗️ **design**] The Hermes reasoning applies to Android too, but `MAX_LOG_STACK_FRAMES` in `crashreporting/CrashReport.kt` is still 25. Can you do the same change there? Otherwise the same crash gets cut at a different depth per platform.

File: packages/expo-app-metrics/ios/CrashReporting/CrashReport.swift [hunk 6]
```diff
@@ -115,10 +121,26 @@ public struct CrashReport: Codable, Sendable {
...
       attributes["expo.crash.objc_exception_message"] = exceptionReason.composedMessage
     }
+    if let callStacks = callStackTree?.callStacks {
+      attributes["expo.crash.thread_count"] = callStacks.count
+      attributes["expo.crash.thread_attributed"] = callStacks.contains { $0.threadAttributed == true }
+    }
+    if let selected, !selected.frames.isEmpty {
+      attributes["expo.crash.rendered_frame_count"] = selected.frames.count
+      attributes["expo.crash.total_frame_count"] = selected.total
+      if let binaryUUIDs = renderedBinaryUUIDs(in: selected.frames) {
```
Comment (jakub-agent) [inline packages/expo-app-metrics/ios/CrashReporting/CrashReport.swift:139]: [💡 suggestion] Can we add a test for this? Every fixture passes `binaryUUID: nil`, so `renderedBinaryUUIDs` always returns `nil` today, and `thread_count` / `thread_attributed` / `rendered_frame_count` / `total_frame_count` aren't asserted either.

File: packages/expo-app-metrics/ios/CrashReporting/CrashReport.swift [hunk 7]
```diff
@@ -187,17 +217,49 @@ public struct CrashReport: Codable, Sendable {
     }
   }
 
-  private func renderFrame(_ frame: CallStackTree.Frame) -> String {
-    if let symbol = frame.symbol {
-      return symbol
+  /// The binaries named in the trace and their UUIDs, as `<binary>:<uuid>` entries.
```
Comment (jakub-agent) [inline packages/expo-app-metrics/ios/CrashReporting/CrashReport.swift:219]: [🐧 nit] nit: the doc comments are getting long across this file — `maxLogStackFrames`, this one, `renderFrame`. Can we cut each to a line or two and leave the longer reasoning in the PR description?

File: packages/expo-app-metrics/ios/Tests/CrashReportTests.swift [hunk 8]
```diff
@@ -266,13 +281,15 @@ struct CrashReportTests {
...
       let lines = stacktrace.split(separator: "\n")
-      #expect(lines.count == 26)
-      #expect(lines.first == "frame0")
-      #expect(lines[24] == "frame24")
+      #expect(lines.count == 51)
+      // These fixtures carry no offset, so only the binary name is appended to each symbol.
+      #expect(lines.first == "frame0 (TestApp)")
+      #expect(lines[49] == "frame49 (TestApp)")
       #expect(lines.last == "… +3 more frames")
       #expect(!stacktrace.contains("unattributed"))
```
Comment (jakub-agent) [inline packages/expo-app-metrics/ios/Tests/CrashReportTests.swift]: [🐧 nit] nit: there's a stray blank line added just below, before the closing brace — can we drop it?

**Review by jakub-agent — APPROVED** [review]
AI Review nits

Otherwise LGTM 👍
